controllers/fenceagentsremediation_controller_test.go

```diff
@@ -200,11 +201,149 @@ var _ = Describe("FAR Controller", func() {
 				By("Still having one test pod")
 				verifyPodExists(testPodName)
 
-				By("Verifying correct conditions for unsuccessfull remediation")
+				By("Verifying correct conditions for unsuccessful remediation")
 				Expect(underTestFAR.Status.LastUpdateTime).ToNot(BeNil())
-				verifyStatusCondition(dummyNode, commonConditions.ProcessingType, conditionStatusPointer(metav1.ConditionFalse))
-				verifyStatusCondition(dummyNode, v1alpha1.FenceAgentActionSucceededType, conditionStatusPointer(metav1.ConditionFalse))
-				verifyStatusCondition(dummyNode, commonConditions.SucceededType, conditionStatusPointer(metav1.ConditionFalse))
+				verifyStatusCondition(underTestFAR.Status.Conditions, dummyNode, commonConditions.ProcessingType, conditionStatusPointer(metav1.ConditionFalse))
+				verifyStatusCondition(underTestFAR.Status.Conditions, dummyNode, utils.FenceAgentActionSucceededType, conditionStatusPointer(metav1.ConditionFalse))
+				verifyStatusCondition(underTestFAR.Status.Conditions, dummyNode, commonConditions.SucceededType, conditionStatusPointer(metav1.ConditionFalse))
+			})
+		})
+
+		Context("Fence Agent Failures", func() {
+			BeforeEach(func() {
+				plogs.Clear()
+				node = utils.GetNode("", workerNode)
+
+				underTestFAR = getFenceAgentsRemediation(workerNode, fenceAgentIPMI, testShareParam, testNodeParam)
+			})
+
+			When("CR is deleted in between fence agent retries", func() {
+				BeforeEach(func() {
+					// Set the controlledRun to fail to simulate a case where CR deletion occurs in between fence agent
+					// command calls.
+					mockError = errors.New("mock error")
+					DeferCleanup(func() { mockError = nil })
+
+					underTestFAR.Spec.RetryCount = 100
+					underTestFAR.Spec.RetryInterval = metav1.Duration{Duration: 1 * time.Second}
+				})
+
+				It("should exit immediately without trying to update the status conditions", func() {
+					Eventually(func(g Gomega) {
+						g.Expect(k8sClient.Get(context.Background(), client.ObjectKey{Name: workerNode}, node)).To(Succeed())
+						g.Expect(utils.TaintExists(node.Spec.Taints, &farNoExecuteTaint)).To(BeTrue(), "remediation taint should exist")
+					}, timeoutFinalizer, pollInterval).Should(Succeed())
+
+					By("Wait some retries")
+					Eventually(func() int {
+						return plogs.CountOccurences("command failed")
+					}, "10s", "1s").Should(BeNumerically(">", 3))
+
+					By("Deleting the CR")
+					Expect(k8sClient.Get(context.Background(), client.ObjectKey{Name: workerNode, Namespace: defaultNamespace}, underTestFAR)).To(Succeed())
+					Expect(k8sClient.Delete(context.Background(), underTestFAR)).To(Succeed())
+
+					By("Verifying goroutine stopped without trying to update the conditions")
+					Eventually(func() bool {
+						return plogs.Contains("fence agent context canceled. Nothing to do")
+					}).Should(BeTrue())
+				})
```

Comment:
This unit test and the next one are quite similar. Can they be under one function which will be called twice for each `when` and `BeforeEach`?

Reply:
The first one needs to _wait for some retries_, while the latter will stuck in its first one. Unfortunately this part is right in the middle two exactly equal pieces of code, but it's decisive.

I noticed a wrong comment in the latter though :smile:

Reply:
>Unfortunately this part is right in the middle two exactly equal pieces of code

Yes, I was thinking of having one function with similar code and only the changing part (the one in the middle) will be decided during runtime based on an input (bool/int/string). It might ease the reading of the unit test logic and easier to detect errors that might impact both of them (since they share similar code).

Reply:
> the changing part (the one in the middle) will be decided during runtime based on an input (bool/int/string)

IMHO this makes the test harder to read. The reader should also figure out the logic of the input and if the changing part has to be done or not

Reply:
ok, I though it would mainly simlify stuff.